File: src/main/java/uknowlab/designpattern/observer/exam1/ObservableInt.java
package uknowlab.designpattern.observer.exam1;

import java.util.ArrayList;
import java.util.List;

public class ObservableInt {
    private int value;

    private final List<Subscriber> subscribers;

    public ObservableInt(int value) {
        this.value = value;
        subscribers = new ArrayList<>();
    }

    public void setValue(int value) {
        this.value = value;
        subscribers.forEach(subscriber -> subscriber.notifyUpdated(value));
    }

    public int getValue() {
        return value;
    }

    public void addSubscriber(Subscriber subscriber) {
        subscribers.add(subscriber);
    }
}

File: src/main/java/uknowlab/designpattern4kotlin/observer/exam1/ObservableInt.kt
package uknowlab.designpattern4kotlin.observer.exam1

import java.util.function.Consumer


class ObservableInt(private var value: Int) {
    private val subscribers: MutableList<Subscriber> =
        ArrayList()

    fun setValue(value: Int) {
        this.value = value
        subscribers.forEach(Consumer { subscriber: Subscriber -> subscriber.notifyUpdated(value) })
    }

    fun getValue(): Int {
        return value
    }

    fun addSubscriber(subscriber: Subscriber) {
        subscribers.add(subscriber)
    }
}
